src/Storages/Distributed/DistributedAsyncInsertBatch.cpp

```diff
@@ -0,0 +1,304 @@
+#include <Storages/Distributed/DistributedAsyncInsertBatch.h>
+#include <Storages/Distributed/DistributedAsyncInsertHelpers.h>
+#include <Storages/Distributed/DistributedAsyncInsertHeader.h>
+#include <Storages/Distributed/DistributedAsyncInsertDirectoryQueue.h>
+#include <Storages/StorageDistributed.h>
+#include <QueryPipeline/RemoteInserter.h>
+#include <Common/CurrentMetrics.h>
+#include <IO/Operators.h>
+#include <IO/WriteBufferFromFile.h>
+
+namespace
+{
+
+namespace fs = std::filesystem;
+
+}
+
+namespace CurrentMetrics
+{
+    extern const Metric DistributedSend;
+}
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int MEMORY_LIMIT_EXCEEDED;
+    extern const int DISTRIBUTED_BROKEN_BATCH_INFO;
+    extern const int DISTRIBUTED_BROKEN_BATCH_FILES;
+    extern const int TOO_MANY_PARTS;
+    extern const int TOO_MANY_BYTES;
+    extern const int TOO_MANY_ROWS_OR_BYTES;
+    extern const int TOO_MANY_PARTITIONS;
+    extern const int DISTRIBUTED_TOO_MANY_PENDING_BYTES;
+    extern const int ARGUMENT_OUT_OF_BOUND;
+}
+
+/// Can the batch be split and send files from batch one-by-one instead?
+bool isSplittableErrorCode(int code, bool remote)
```

Comment:
Well it is not used anywhere else, except for batch INSERT, since only for it `split_batch_on_failure` make sense, and unlikely it will be needed by another, new, code.
So I think that it is better to keep it in one file.